### PR TITLE
[SPARK-30112][SQL] Allow insert overwrite same table if using dynamic partition overwrite

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch proposes to allow insert overwrite same table if using dynamic partition overwrite.

### Why are the changes needed?

Currently, Insert overwrite cannot overwrite to same table even it is dynamic partition overwrite. But for dynamic partition overwrite, we do not delete partition directories ahead. We write to staging directories and move data to final partition directories. We should be able to insert overwrite to same table under dynamic partition overwrite.

This enables users to read data from a table and insert overwrite to same table by using dynamic partition overwrite. Because this is not allowed for now, users need to write to other temporary location and move it back to the table.

### Does this PR introduce any user-facing change?

Yes. Users can insert overwrite same table if using dynamic partition overwrite.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @cloud-fan

**Comment by viirya** [comment]
retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Thank you for making a PR for this issue.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 1]
```diff
@@ -270,6 +270,44 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
       "INSERT OVERWRITE to a table while querying it should not be allowed.")
   }
 
+  test("it is allowed to write to a table while querying it for dynamic partition overwrite.") {
+    Seq(PartitionOverwriteMode.DYNAMIC.toString,
+        PartitionOverwriteMode.STATIC.toString).foreach { usedMode =>
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: `usedMode` -> `mode`.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 2]
```diff
@@ -270,6 +270,44 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
...
+      withSQLConf(SQLConf.PARTITION_OVERWRITE_MODE.key -> usedMode) {
+        withTable("insertTable") {
+          sql(
+            """
+              |CREATE TABLE insertTable(i int, part1 int, part2 int) USING PARQUET
+              |PARTITIONED BY (part1, part2)
+           """.stripMargin)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: indentation?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 3]
```diff
@@ -270,6 +270,44 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
...
+                |SELECT i + 1, part2 FROM insertTable
+              """.stripMargin)
+            checkAnswer(spark.table("insertTable"), Row(2, 1, 1))
+          } else {
+            val message = intercept[AnalysisException] {
+              sql(
+                """
+                  |INSERT OVERWRITE TABLE insertTable PARTITION(part1=1, part2)
+                  |SELECT i + 1, part2 FROM insertTable
+              """.stripMargin)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: indentation?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala (hunk 2 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: oops! thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 4]
```diff
@@ -270,6 +270,44 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
...
+      withSQLConf(SQLConf.PARTITION_OVERWRITE_MODE.key -> usedMode) {
+        withTable("insertTable") {
+          sql(
+            """
+              |CREATE TABLE insertTable(i int, part1 int, part2 int) USING PARQUET
+              |PARTITIONED BY (part1, part2)
+           """.stripMargin)
+
+          sql("INSERT INTO TABLE insertTable PARTITION(part1=1, part2=1) SELECT 1")
+          checkAnswer(spark.table("insertTable"), Row(1, 1, 1))
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: Shall we have a test table with multiple partitions with data?

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: good for that. Will add one.

**Comment by dongjoon-hyun** [comment]
cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala [hunk 5]
```diff
@@ -60,6 +61,21 @@ case class InsertIntoHadoopFsRelationCommand(
   extends DataWritingCommand {
   import org.apache.spark.sql.catalyst.catalog.ExternalCatalogUtils.escapePathName
 
+  private lazy val parameters = CaseInsensitiveMap(options)
+
+  private[sql] lazy val dynamicPartitionOverwriteEnabled: Boolean = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala]: shall we just name it `dynamicPartitionOverwrite`? then we can remove https://github.com/apache/spark/pull/26752/files#diff-f9858c2d9d1a3c3e48753ef675bc865aR109

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 6]
```diff
@@ -270,6 +270,54 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
       "INSERT OVERWRITE to a table while querying it should not be allowed.")
   }
 
+  test("it is allowed to write to a table while querying it for dynamic partition overwrite.") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: shall we put the JIRA id?

**Review by cloud-fan — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master.
Thank you so much, @viirya and @cloud-fan .
I also verified the last commit manually.

**Comment by viirya** [comment]
thanks! @cloud-fan @dongjoon-hyun

**Pull request closed without merging**